pages/DSNP/UserData.md

```diff
@@ -0,0 +1,211 @@
+
+An `etag` is any ASCII-encoded string with a minimum length of one byte and a maximum length of 255 bytes.
+
+An implementation MUST fail the Replace User Data Operation if the `etag` values supplied do not match the current `etag` values for all chunks of the specified data type.
+
+## Replace User Data Operation
+
+The Replace User Data Operation takes the following parameters:
+
+* A DSNP User Id
+  * Implementations MUST ensure that the principal invoking this Operation is this user, or a transparent chain of delegation from the user to the principal exists.
+* A [Key Identifier](Types/PublicKey.md#keyid) for the `keyAgreement` key pair used to encrypt any private data in the operation.
+  (If only unencrypted user data types are included, the key identifier is optional.)
+* A map containing the set of data types to update as the keys, and tuples consisting of (1) the schema version used to encode the data type, and (2) a list where each element includes a data chunk and its associated entity tag, as the values.
+
+If the Operation is successful, any previous data associated with the user for each data type included in the input MUST be removed and replaced by the new data.
+
+Data chunks should be generated for each included data type using the following sequence of operations:
+1. Serialize the data records, in Avro binary format, according to the versioned schema.
+1. Divide the data records into one or more chunks, with each chunk containing a maximum byte count specified by the implementation.
+* Note that the combined chunks constitute a different, and smaller, binary than a complete Avro file, as they exclude the header information, schema, and any block-level metadata.
+  Each chunk should start and end with a complete Avro object; in other words, chunk boundaries MUST NOT occur mid-record.
+    Applications SHOULD try to include as many records as possible within the byte length limit, but because compression and encryption may alter the byte size versus the raw Avro binary records, they may wish to use a heuristic approach that sometimes produces non-optimal chunks.
+3. For each chunk generated, the application should then:
+    1. If the data type requires compression, apply the compression codec noted.
+    1. If the data type requires encryption,
+        1. Retrieve the user's active (most recently announced) `keyAgreement` public key, U<sub>public</sub>.
+  The `keyId` in the announcement should match the key identifier provided for this Operation.
+  If no key exists, one should be created and published as an Announcement before invoking the Operation.
+        1. Create a sealed box, as in the [libsodium](https://doc.libsodium.org/public-key_cryptography/sealed_boxes) function `crypto_box_seal`, using U<sub>public</sub>.
```

Comment:
so seal box would need creation of an ephemeral key, do we need to include that or it is already assumed since we are using `crypto_box_seal`? 

On another topic do we need another layer of key derivation with different context for encryption or is that an overkill?

Reply:
I added more description for "sealed box", hopefully that helps.

For the second comment, if I'm understanding it correctly, yes, we will add an additional context for PRIds used for message encryption (just not part of this PR).